File: cli/src/generate.ts
import path from "node:path";
import fs from "node:fs/promises";
import Handlebars from "handlebars";
import { pathExists, insertAtAnchorSafe } from "./fsutil.ts";
import type { FieldSpec, FieldType } from "./nlp.ts";

Handlebars.registerHelper("pascal", (s: string) => s[0].toUpperCase() + s.slice(1));
Handlebars.registerHelper("capitalize", (s: string) => s[0].toUpperCase() + s.slice(1));
Handlebars.registerHelper("eq", (a: unknown, b: unknown) => a === b);
Handlebars.registerHelper("ne", (a: unknown, b: unknown) => a !== b);

interface RenderedField {
  name: string;
  last: boolean;
  // TS bases
  tsType?: string;
  zodType?: string;
  drizzleField?: string;
  // Python bases
  pyType?: string;
  sqlType?: string;
}

const TS_TYPE: Record<FieldType, string> = {
  string: "string",
  number: "number",
  boolean: "boolean",
  date: "string",
  relation: "string",
};
const ZOD_TYPE: Record<FieldType, string> = {
  string: "z.string()",
  number: "z.number()",
  boolean: "z.boolean()",
  date: "z.string()",
  relation: "z.string()",
};
const PY_TYPE: Record<FieldType, string> = {
  string: "str",
  number: "float",
  boolean: "bool",
  date: "str",
  relation: "str",
};
const SQL_TYPE: Record<FieldType, string> = {
  string: "TEXT",
  number: "REAL",
  boolean: "INTEGER",
  date: "TEXT",
  relation: "TEXT",
};

function drizzleColumn(name: string, type: FieldType): string {
  switch (type) {
    case "number":
      return `${name}: real("${name}")`;
    case "boolean":
      return `${name}: integer("${name}", { mode: "boolean" })`;
    case "date":
      return `${name}: integer("${name}", { mode: "timestamp" })`;
    default:
      return `${name}: text("${name}")`;
  }
}

interface ScaffoldWiring {
  file: string;
  anchor: string;
  insert: string;
}

interface ScaffoldConfig {
  language: "ts" | "python";
  templates: Record<string, string>; // template filename -> target path pattern, both may use {{entityName}}
  wiring: (ScaffoldWiring & { optional?: boolean })[];
}

export interface GeneratePageOptions {
  registryPath: string;
  targetDir: string;
  base: string;
  entityName: string;
  fields: FieldSpec[];
  auth: "protected" | "public";
}

export interface GeneratePageResult {
  filesWritten: string[];
  wiringApplied: string[];
  wiringSkipped: string[];
}

function substitute(pattern: string, entityName: string, entityNamePascal: string): string {
  return pattern.replaceAll("{{entityName}}", entityName).replaceAll("{{entityNamePascal}}", entityNamePascal);
}

export async function generatePage(opts: GeneratePageOptions): Promise<GeneratePageResult> {
  const { registryPath, targetDir, base, entityName, fields, auth } = opts;
  const scaffoldDir = path.join(registryPath, "pages", "_scaffold", base);
  const configPath = path.join(scaffoldDir, "scaffold.config.json");
  if (!(await pathExists(configPath))) {
    throw new Error(`No page-add scaffold available for base "${base}"`);
  }
  const config: ScaffoldConfig = JSON.parse(await fs.readFile(configPath, "utf-8"));
  const entityNamePascal = entityName[0].toUpperCase() + entityName.slice(1);

  const renderedFields: RenderedField[] = fields.map((f, i) => {
    const base: RenderedField = { name: f.name, last: i === fields.length - 1 };
    if (config.language === "python") {
      base.pyType = PY_TYPE[f.type];
      base.sqlType = SQL_TYPE[f.type];
    } else {
      base.tsType = TS_TYPE[f.type];
      base.zodType = ZOD_TYPE[f.type];
      base.drizzleField = drizzleColumn(f.name, f.type);
    }
    return base;
  });

  const vars = {
    entityName,
    entityNamePascal,
    fields: renderedFields,
    protected: auth === "protected",
  };

  const filesWritten: string[] = [];
  for (const [tpl, targetPattern] of Object.entries(config.templates)) {
    const tplPath = path.join(scaffoldDir, tpl);
    if (!(await pathExists(tplPath))) continue;
    const raw = await fs.readFile(tplPath, "utf-8");
    const rendered = Handlebars.compile(raw)(vars);
    const relTarget = substitute(targetPattern, entityName, entityNamePascal);
    const outPath = path.join(targetDir, relTarget);
    await fs.mkdir(path.dirname(outPath), { recursive: true });
    await fs.writeFile(outPath, rendered);
    filesWritten.push(relTarget);
  }

  const wiringApplied: string[] = [];
  const wiringSkipped: string[] = [];
  for (const w of config.wiring ?? []) {
    const file = substitute(w.file, entityName, entityNamePascal);
    const insert = substitute(w.insert, entityName, entityNamePascal);
    const applied = await insertAtAnchorSafe(path.join(targetDir, file), w.anchor, insert);
    if (applied) wiringApplied.push(file);
    else wiringSkipped.push(`${file} (feature that owns this anchor isn't installed — wire it manually)`);
  }

  return { filesWritten, wiringApplied, wiringSkipped };
}

File: go-cli/internal/generate/generate.go
// Package generate ports cli/src/generate.ts: the `hackpack page add`
// CRUD-scaffold generator, driven by raymond (Handlebars) for file bodies and
// plain string substitution (not Handlebars) for target paths and wiring
// file/insert strings — matching the TS implementation's `substitute()`,
// which is a literal .replaceAll, not a Handlebars.compile call.
package generate

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/aymerick/raymond"

	"hackpack/internal/fsutil"
)

func init() {
	raymond.RegisterHelper("pascal", func(s string) string {
		if s == "" {
			return s
		}
		return strings.ToUpper(s[:1]) + s[1:]
	})
}

type FieldType string

const (
	FieldString   FieldType = "string"
	FieldNumber   FieldType = "number"
	FieldBoolean  FieldType = "boolean"
	FieldDate     FieldType = "date"
	FieldRelation FieldType = "relation"
)

var ValidFieldTypes = []FieldType{FieldString, FieldNumber, FieldBoolean, FieldDate, FieldRelation}

type FieldSpec struct {
	Name string
	Type FieldType
}

var tsType = map[FieldType]string{
	FieldString: "string", FieldNumber: "number", FieldBoolean: "boolean",
	FieldDate: "string", FieldRelation: "string",
}
var zodType = map[FieldType]string{
	FieldString: "z.string()", FieldNumber: "z.number()", FieldBoolean: "z.boolean()",
	FieldDate: "z.string()", FieldRelation: "z.string()",
}
var pyType = map[FieldType]string{
	FieldString: "str", FieldNumber: "float", FieldBoolean: "bool",
	FieldDate: "str", FieldRelation: "str",
}
var sqlType = map[FieldType]string{
	FieldString: "TEXT", FieldNumber: "REAL", FieldBoolean: "INTEGER",
	FieldDate: "TEXT", FieldRelation: "TEXT",
}

func drizzleColumn(name string, t FieldType) string {
	switch t {
	case FieldNumber:
		return fmt.Sprintf(`%s: real("%s")`, name, name)
	case FieldBoolean:
		return fmt.Sprintf(`%s: integer("%s", { mode: "boolean" })`, name, name)
	case FieldDate:
		return fmt.Sprintf(`%s: integer("%s", { mode: "timestamp" })`, name, name)
	default:
		return fmt.Sprintf(`%s: text("%s")`, name, name)
	}
}

type scaffoldWiring struct {
	File     string `json:"file"`
	Anchor   string `json:"anchor"`
	Insert   string `json:"insert"`
	Optional bool   `json:"optional,omitempty"`
}

type scaffoldConfig struct {
	Language  string            `json:"language"`
	Templates map[string]string `json:"templates"`
	Wiring    []scaffoldWiring  `json:"wiring"`
}

type Options struct {
	RegistryPath string
	TargetDir    string
	Base         string
	EntityName   string
	Fields       []FieldSpec
	Auth         string // "protected" | "public"
}

type Result struct {
	FilesWritten  []string
	WiringApplied []string
	WiringSkipped []string
}

func substitute(pattern, entityName, entityNamePascal string) string {
	pattern = strings.ReplaceAll(pattern, "{{entityName}}", entityName)
	pattern = strings.ReplaceAll(pattern, "{{entityNamePascal}}", entityNamePascal)
	return pattern
}

func GeneratePage(opts Options) (Result, error) {
	scaffoldDir := filepath.Join(opts.RegistryPath, "pages", "_scaffold", opts.Base)
	configPath := filepath.Join(scaffoldDir, "scaffold.config.json")
	if !fsutil.PathExists(configPath) {
		return Result{}, fmt.Errorf("no page-add scaffold available for base %q", opts.Base)
	}
	raw, err := os.ReadFile(configPath)
	if err != nil {
		return Result{}, err
	}
	var cfg scaffoldConfig
	if err := json.Unmarshal(raw, &cfg); err != nil {
		return Result{}, err
	}

	entityName := opts.EntityName
	entityNamePascal := strings.ToUpper(entityName[:1]) + entityName[1:]

	renderedFields := make([]map[string]interface{}, len(opts.Fields))
	for i, f := range opts.Fields {
		rf := map[string]interface{}{
			"name": f.Name,
			"last": i == len(opts.Fields)-1,
		}
		if cfg.Language == "python" {
			rf["pyType"] = pyType[f.Type]
			rf["sqlType"] = sqlType[f.Type]
		} else {
			rf["tsType"] = tsType[f.Type]
			rf["zodType"] = zodType[f.Type]
			rf["drizzleField"] = drizzleColumn(f.Name, f.Type)
		}
		renderedFields[i] = rf
	}

	vars := map[string]interface{}{
		"entityName":       entityName,
		"entityNamePascal": entityNamePascal,
		"fields":           renderedFields,
		"protected":        opts.Auth == "protected",
	}

	var filesWritten []string
	for tpl, targetPattern := range cfg.Templates {
		tplPath := filepath.Join(scaffoldDir, tpl)
		if !fsutil.PathExists(tplPath) {
			continue
		}
		rawTpl, err := os.ReadFile(tplPath)
		if err != nil {
			return Result{}, err
		}
		rendered, err := raymond.Render(string(rawTpl), vars)
		if err != nil {
			return Result{}, fmt.Errorf("rendering %s: %w", tplPath, err)
		}
		relTarget := substitute(targetPattern, entityName, entityNamePascal)
		outPath := filepath.Join(opts.TargetDir, relTarget)
		if err := os.MkdirAll(filepath.Dir(outPath), 0755); err != nil {
			return Result{}, err
		}
		if err := os.WriteFile(outPath, []byte(rendered), 0644); err != nil {
			return Result{}, err
		}
		filesWritten = append(filesWritten, relTarget)
	}

	var wiringApplied, wiringSkipped []string
	for _, w := range cfg.Wiring {
		file := substitute(w.File, entityName, entityNamePascal)
		insert := substitute(w.Insert, entityName, entityNamePascal)
		if fsutil.InsertAtAnchorSafe(filepath.Join(opts.TargetDir, file), w.Anchor, insert) {
			wiringApplied = append(wiringApplied, file)
		} else {
			wiringSkipped = append(wiringSkipped, fmt.Sprintf("%s (feature that owns this anchor isn't installed — wire it manually)", file))
		}
	}

	return Result{FilesWritten: filesWritten, WiringApplied: wiringApplied, WiringSkipped: wiringSkipped}, nil
}
